lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1172 @@
+
+	// This helps avoiding doing the same check in the visitBranch method
+
+	// What we need to do here is something very important: 
+	// check the operation in mostly the same way 
+	// but now I do not need to look for the constant but for the presence of the rlc.uninit_construct
+	
+	// Before this there could be a case in which I have only one successor, 
+	// in that case I just call the visitYield
+	if(this->getOperation()->getBlock()->getNumSuccessors()==1)
+		return mlir::rlc::visitYield(this->getOperation(), currentLattice);
+
+	mlir::Operation* conditional_op=casted->getOperand(1).getDefiningOp();
+
+	// Do the checking only if the operation is a conditional (relational)
+	if(	not mlir::isa<mlir::rlc::LessOp>(conditional_op) and 
+		not mlir::isa<mlir::rlc::GreaterOp>(conditional_op) and
+		not mlir::isa<mlir::rlc::LessEqualOp>(conditional_op) and 
+		not mlir::isa<mlir::rlc::GreaterEqualOp>(conditional_op))
+		return false;
+
+	// Aliases
+	const auto& lhs = conditional_op->getOperand(0);
+	const auto& rhs = conditional_op->getOperand(1);
+	const auto& lhs_op = lhs.getDefiningOp();
+	const auto& rhs_op = rhs.getDefiningOp();
+
+	// The algorithm will work mostly as the one for the arithmetic operations, but it will be simpler (no commutativity problem)
+	// NB: note that they are not exclusive so they have to be executed always
+
+	return currentLattice->updateRangeFromVisit(this->getOperation(), analysis, conditional_op, rhs_op, lhs_op, rhs, lhs);
+
+}
+
+namespace mlir::rlc
+{
+	//////////////////////////////////////////////////////////////////////////////
+	// USEFUL STATIC METHODS
+	//////////////////////////////////////////////////////////////////////////////
+
+	static bool findInterestingOperandAndUpdate(
+		mlir::Operation* op,
+		mlir::rlc::ConstraintsLattice* currentLattice,
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result){
+
+		// Keep operand aliases for clarity
+		const auto& op1=op->getOperand(0);
+		const auto& op2=op->getOperand(1);
+		const auto& op1_defop=op1.getDefiningOp();
+		const auto& op2_defop=op2.getDefiningOp();
+
+		// Understand the operand that I have to update
+		mlir::Value update=nullptr;
+		mlir::Value  other=nullptr;
+
+		// Check which operand should be updated (the other should be a constant)
+		// TODO: understand what to do when there are no constants
+		if(op1_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op1_defop)){
+			update=op1;
+			other=op2;
+		}
+		else if(op2_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op2_defop)){
+			update=op2;
+			other=op1;
+		}
+		else return false;
+
+		bool ret1=mlir::rlc::updateArithmeticRange(op, currentLattice, analysis, result, update, other, true);
+		bool ret2=mlir::rlc::updateArithmeticRange(op, currentLattice, analysis, result, update, other, false);
+
+		return ret1 or ret2;
+	}
+
+	// Method that checks the lattice and updates the value accordingly
+	static bool updateArithmeticRange(
```

Comment:
i understand that this is tied to the previous free function, since it is used only there, can we get a better name for this guy too?